compiler/riscv/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2330,19 +2344,61 @@ TR::Instruction *loadConstant64(TR::CodeGenerator *cg, TR::Node *node, int64_t v
       }
    else
       {
-      int32_t lo32 = (int32_t)((uint64_t)value & 0xFFFFFFFF);
-      int32_t hi32 = (int32_t)((uint64_t)value >> 32);
-      if (lo32 < 0)
+      /*
+       * For truly 64bit values, we first load *high* 32bits of the value into
+       * target registers' *low* 32bits and then use a sequence of slli + addi
+       * instructions to load lower 32 bits.
+       *
+       * We load only 11bits at the time making sure the highest bit of
+       * 12bit immediate is zero - this is to overcome the fact that I-type
+       * instructions *sign-extend* the immediate value before performing the
+       * operation.
+       */
+      uint32_t hi32 = (uint32_t)((uint64_t)value >> 32);
+
+      const uint32_t nbits = RISCV_IMM_BITS - 1; // how much bits to add at each step
+      uint32_t toShift = 0; // how much to shift left before adding lower bits of the value
+      uint32_t bits = 0; // actual bits to add add each step
+
+      cursor = loadConstant32(cg, node, hi32, trgReg, cursor);
+
+       /*
+       * Add high 11 bits of the low 32 bits
+       */
+      bits = extractBits(value, 31 - 0*nbits, 31 - 1*nbits + 1);
+      toShift += hi32 == 0 ? 0 : nbits;
+      if (bits)
+         {
+         if (toShift)
+            cursor = generateITYPE(TR::InstOpCode::_slli, node, trgReg, trgReg, toShift, cg);
```

Comment:
I believe you want to use the form that accepts the previous instruction (i.e., `cursor`), otherwise this will append to the end of the instruction stream.

This same comment applies to the remaining "generateXXX" calls in this function.